test/restart.jl

```diff
@@ -0,0 +1,338 @@
+# nested testsets have to be using let blocks.
+#
+# We need to try-catch because @test thorws exceptions. Exceptions are normally captured by
+# the @testset, but this does not work when using a let block. :(
+#
+# And there's a final snag: by try-catching, the error is not propagated, so the
+# job is not marked as failing. We have to manually trigger exit with an error
+# status when the job has failed.
+const SUCCESS::Base.RefValue{Bool} = Ref(true)
+
+"""
+    _error(arr1::AbstractArray, arr2::AbstractArray; ABS_TOL = 100eps(eltype(arr1)))
+
+We compute the error in this way:
+- when the absolute value is larger than ABS_TOL, we use the absolute error
+- in the other cases, we compare the relative errors
+"""
+function _error(
+    arr1::AbstractArray,
+    arr2::AbstractArray;
+    ABS_TOL = 100eps(eltype(arr1)),
+)
+    diff = abs.(arr1 .- arr2)
+    denominator = abs.(arr1)
+    error = ifelse.(denominator .> ABS_TOL, diff ./ denominator, diff)
+    return error
+end
+
+function rmap(val; ignore = Set([:rc]))
+    properties = filter(x -> !(x in ignore), propertynames(val))
+    properties = propertynames(val)
+end
+
+"""
+    compare(v1, v2; name = "", ignore = Set([:rc]))
+
+Recursively compare `v1` and `v2` up to some numeric tolerance.
+
+`compare` walks through all the properties in `v1` and `v2` until it finds
+that there are no more properties. At that point, `compare` tries to match the
+resulting objects. When such objects are arrays with floating point, `compare`
+defines a notion of `error` that is the following: when the absolute value is
+less than `100eps(eltype)`, `error = absolute_error`, otherwise it is relative
+error. The `error` is then compared against a tolerance.
+
+Keyword arguments
+=================
+
+- `name` is used to collect the name of the property while we go recursively
+  over all the properties. You can pass a base name.
+- `ignore` is a collection of `Symbol`s that identify properties that are
+  ignored when walking through the tree. This is useful for properties that
+  are known to be different (e.g., `output_dir`).
+
+`:rc` is some CUDA/CuArray internal object that we don't care about
+"""
+function compare(v1, v2; name = "", ignore = Set([:rc]))
+    pass = true
+    return _compare(pass, v1, v2; name, ignore)
+end
+
+function _compare(pass, v1::T, v2::T; name, ignore) where {T}
+    properties = filter(x -> !(x in ignore), propertynames(v1))
+    if isempty(properties)
+        # We need to handle the case with a Field with only one variable
+        v1 isa Field && (v1 = parent(v1); v2 = parent(v2))
+        pass &= _compare(v1, v2; name, ignore)
+    else
+        # Recursive case
+        for p in properties
+            pass &= _compare(
+                pass,
+                getproperty(v1, p),
+                getproperty(v2, p);
+                name = "$(name).$(p)",
+                ignore,
+            )
+        end
+    end
+    return pass
+end
```

Comment:
```suggestion
function _compare(
    pass,
    v1::T,
    v2::T;
    name,
    ignore,
) where {T <: Fields.Field}
    return pass && _compare(pass, Fields.field_values(v1), Fields.field_values(v2); name, ignore)
end
function _compare(pass, v1::T, v2::T; name, ignore) where {T}
    # Recursive case
    for p in filter(x -> !(x in ignore), propertynames(v1))
        pass &= _compare(
            pass,
            getproperty(v1, p),
            getproperty(v2, p);
            name = "$(name).$(p)",
            ignore,
        )
    end
    return pass
end
```
Can't we break down this last leaf case so that the generic recursive case is independent of ClimaCore?

Reply:
This doesn't walk into the properties of the Field when Fields have multiple properties (e.g. `u.c`). We want to display difference one variable at the time because that's what's the most informative.